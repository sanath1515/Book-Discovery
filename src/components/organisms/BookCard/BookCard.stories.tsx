import React from "react";
import BookCard from "./BookCard";

export default {
  title: "Organisms/BookCard",
  component: BookCard,
};

const Template = () => (
  <BookCard
    title="Nature Chemistry"
    desc="Nature chemistry deals with different biocatalytic approaches to transform phenols by adding different neurons"
    authorName="J D Lee"
    image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png"
  />
);

export const Default = Template.bind({});
