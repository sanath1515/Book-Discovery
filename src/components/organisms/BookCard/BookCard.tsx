import React, { useState } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import BookmarkIcon from "@material-ui/icons/Bookmark";
import PlaylistAddCheckIcon from "@material-ui/icons/PlaylistAddCheck";
import Rating from "@material-ui/lab/Rating";
import ReplySharpIcon from "@material-ui/icons/ReplySharp";
import BookmarkBorderIcon from "@material-ui/icons/BookmarkBorder";
import { CardActions, Grid, Typography } from "@material-ui/core";
import { COLORS } from "../../../theme/constants";
import Avatar from "../../atoms/Avatar";
import theme from "../../../theme/theme";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router";

const useStyles = makeStyles({
  root: {
    width: "530px",
    height: "255px",
    padding: theme.spacing(5, 2, 4, 3),
    borderRadius: "4px",
    border: "solid 1px",
    borderColor: COLORS.GREY_300,
		"& .MuiPaper-elevation1":{
			boxShadow: "none"
		}
  },
  image: {
    width: "180px",
    height: "250px",
    margin: theme.spacing(0, 1),
    borderRadius: "4px",
    objectFit: "initial",
    cursor :"pointer"
  },
  title: {
    width: "304px",
    height: "30px",
    margin: theme.spacing(0, 1.2),
    color: COLORS.CHARCOAL_GREY,
  },
  desc: {
    width: "290px",
    height: "66px",
    margin: theme.spacing(2),
    color: COLORS.CHARCOAL_GREY_100,
    lineHeight: "22px",
  },
  avatar: {
    width: "38px",
    height: "38px",
    margin: theme.spacing(2, 2),
  },
  author: {
    width: "36px",
    height: "18px",
    textTransform: "none",
    color: COLORS.ICON_500,
    margin:theme.spacing(2,0)
  },
  authorName: {
    width: "auto",
    height: "24px",
    color: COLORS.CHARCOAL_GREY,
  },
  star: {
    width: "20px",
    height: "22px",
    margin: theme.spacing(1.5, 1),
    color: COLORS.CHARCOAL_GREY,
  },
  rating: {
    width: "98px",
    height: "22px",
    padding: theme.spacing(2, 0),
    color: COLORS.CHARCOAL_GREY,
  },
  bookmark: {
    color: COLORS.PRIMARY_COLOR_500,
    paddingRight: "20px",
    cursor: "pointer"
  },
  bookmarkBorder: {
    color: COLORS.ICON_500,
    paddingRight: "20px",
    cursor: "pointer"
  },
  reply: {
    color: COLORS.GREY_500,
    paddingRight: "20px",
  },
  playlist: {
    color: COLORS.GREY_500,
  },
  action: {
    display: "flex",
    width: "auto",
    padding : theme.spacing(0)
  },
  content: {
    padding: theme.spacing(0),
  },
  author1: {
    marginTop: theme.spacing(1.2),
  },
});
export type BookCardProps = {
  id: string;
  title: string;
  desc: string;
  authorName: string;
  image: string;
  authorImg:string;
};
const BookCard: React.FC<BookCardProps> = ({
  id,
  title,
  desc,
  authorName,
  image,
  authorImg
}) => {
  const dispatch = useDispatch();
  const history = useHistory();

  const handleClick = (bookId: string) => {
    dispatch({ type: "SET_SINGLE_VIEW_ID", payload: bookId });
    history.push("/bookdetail");
  };
  const classes = useStyles();
  const [bookMark, setBookMark] = useState(false);
  const onBookmarkClick = () => {
    setBookMark(!bookMark);
  };
  return (
    <Card className={classes.root}>
      <CardContent className={classes.action}>
        <CardMedia
          className={classes.image}
          component="img"
          image={image}
          onClick={()=>handleClick(id)} data-testid="card"/>
        <CardContent className={classes.content}>
          <Typography className={classes.title} variant="h4">
            {title}
          </Typography>
          <Typography className={classes.desc} variant="body2">
            {desc}
          </Typography>
          <Grid direction="row" container>
            <Grid item>
              <Avatar className={classes.avatar} src={authorImg} />
            </Grid>
            <Grid item className={classes.author1}>
              <Typography className={classes.author} variant="overline">
                Author
              </Typography>
              <Typography className={classes.authorName} variant="subtitle2">
                {authorName}
              </Typography>
            </Grid>
          </Grid>
          <CardActions>
            <Rating
              name="half-rating-read"
              defaultValue={4.5}
              precision={0.5}
              size="small"
              readOnly
            />
            <Typography className={classes.star} variant="body1">
              4.5
            </Typography>
            <Typography className={classes.rating} variant="body2">
              ( 1,204 ratings )
            </Typography>
          </CardActions>
        
      <CardActions>
        {bookMark && (
          <BookmarkIcon
            className={classes.bookmark}
            onClick={onBookmarkClick}
            data-testid="bookmark"
          />
        )}
        {!bookMark && (
          <BookmarkBorderIcon
            className={classes.bookmarkBorder}
            onClick={onBookmarkClick}
            data-testid="bookmark"
          />
        )}
        <ReplySharpIcon className={classes.reply} />
        <PlaylistAddCheckIcon className={classes.playlist} />
      </CardActions>
      </CardContent>
      </CardContent>
    </Card>
  );
};

export default BookCard;
