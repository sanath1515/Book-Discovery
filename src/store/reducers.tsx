export type bookState = {
    searchText:string,
    category:string,
    singleViewId:string
}

export const initialBookState = {
    searchText:"",
    category:"",
    singleViewId:"1"
}

type Action = {type:string,payload:string}

export const booksReducer = (state:bookState = initialBookState,action:Action)=>
{
    switch(action.type)
    {
        case "SET_CATEGORY":{
            return {...state,category:action.payload};
        }
        case "SET_SEARCH_TEXT":{
            return {...state,searchText:action.payload};
        }
        case "SET_SINGLE_VIEW_ID":{
            return {...state,singleViewId:action.payload};
        }
        default:
            return state;
    }
}