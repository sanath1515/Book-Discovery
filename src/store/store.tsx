import { createStore } from 'redux';
import { booksReducer } from './reducers';

const store = createStore(booksReducer);

export default store;