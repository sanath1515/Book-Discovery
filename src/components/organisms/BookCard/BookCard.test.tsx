import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import BookCard from "./BookCard";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import store from "../../../store/store";

afterEach(cleanup);

describe("BookCard Testing", () => {
  it("Matches Snapshot", () => {
    const { container } = render(
      <Router>
      <Provider store={store}>
      <BookCard
            title="Nature Chemistry"
            desc="Nature chemistry deals with different biocatalytic approaches to transform phenols by adding different neurons"
            authorName="J D Lee"
            image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png" id={""} authorImg={""}      />
      </Provider>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("Checking Fields", () => {
    render(
      <Router>
      <Provider store={store}>
      <BookCard
            title="Nature Chemistry"
            desc="Nature chemistry deals with different biocatalytic approaches to transform phenols by adding different neurons"
            authorName="J D Lee"
            image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png" id={""} authorImg={""}      />
      </Provider>
      </Router>
    );

    expect(screen.getByText("Nature Chemistry")).toBeInTheDocument();
    expect(screen.getByText("J D Lee")).toBeInTheDocument();
    expect(
      screen.getByText(
        "Nature chemistry deals with different biocatalytic approaches to transform phenols by adding different neurons"
      )
    ).toBeInTheDocument();
  });
  it("Should click bookmark correctly", () => {
    const { getByTestId } = render(
      <Router>
        <Provider store={store}>
        <BookCard
            title="Nature Chemistry"
            desc="Nature chemistry deals with different biocatalytic approaches to transform phenols by adding different neurons"
            authorName="J D Lee"
            image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png" id={""} authorImg={""}      />
        </Provider>
      </Router>
    );
    fireEvent.click(getByTestId("bookmark"));
    expect(fireEvent.click(getByTestId("bookmark"))).toBeTruthy();
  });
  it("Should click correctly", () => {
    const { getByTestId } = render(
      <Router>
        <Provider store={store}>
        <BookCard
            title="Nature Chemistry"
            desc="Nature chemistry deals with different biocatalytic approaches to transform phenols by adding different neurons"
            authorName="J D Lee"
            image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png" id={""} authorImg={""}      />
        </Provider>
      </Router>
    );
    fireEvent.click(getByTestId("card"));
    expect(fireEvent.click(getByTestId("card"))).toBeTruthy();
  });
});
