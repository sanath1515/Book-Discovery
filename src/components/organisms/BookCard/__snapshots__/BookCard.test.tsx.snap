// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard Testing Matches Snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root makeStyles-action-14"
    >
      <img
        class="MuiCardMedia-root makeStyles-image-2 MuiCardMedia-media MuiCardMedia-img"
        data-testid="card"
        src="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png"
      />
      <div
        class="MuiCardContent-root makeStyles-content-15"
      >
        <h4
          class="MuiTypography-root makeStyles-title-3 MuiTypography-h4"
        >
          Nature Chemistry
        </h4>
        <p
          class="MuiTypography-root makeStyles-desc-4 MuiTypography-body2"
        >
          Nature chemistry deals with different biocatalytic approaches to transform phenols by adding different neurons
        </p>
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular makeStyles-avatar-5 MuiAvatar-colorDefault"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiAvatar-fallback"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiGrid-root makeStyles-author1-16 MuiGrid-item"
          >
            <span
              class="MuiTypography-root makeStyles-author-6 MuiTypography-overline"
            >
              Author
            </span>
            <h6
              class="MuiTypography-root makeStyles-authorName-7 MuiTypography-subtitle2"
            >
              J D Lee
            </h6>
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing"
        >
          <span
            aria-label="4.5 Stars"
            class="MuiRating-root MuiRating-sizeSmall MuiRating-readOnly"
            role="img"
          >
            <span
              class="MuiRating-decimal"
            >
              <span
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
            </span>
            <span
              class="MuiRating-decimal"
            >
              <span
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
            </span>
            <span
              class="MuiRating-decimal"
            >
              <span
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
            </span>
            <span
              class="MuiRating-decimal"
            >
              <span
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
            </span>
            <span
              class="MuiRating-decimal"
            >
              <span
                style="width: 50%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
              <span>
                <span
                  class="MuiRating-icon MuiRating-iconEmpty"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
              </span>
            </span>
          </span>
          <p
            class="MuiTypography-root makeStyles-star-8 MuiTypography-body1"
          >
            4.5
          </p>
          <p
            class="MuiTypography-root makeStyles-rating-9 MuiTypography-body2"
          >
            ( 1,204 ratings )
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-bookmarkBorder-11"
            data-testid="bookmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15l-5-2.18L7 18V5h10v13z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-reply-12"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 9V5l-7 7 7 7v-4.1c5 0 8.5 1.6 11 5.1-1-5-4-10-11-11z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-playlist-13"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 10H2v2h12v-2zm0-4H2v2h12V6zM2 16h8v-2H2v2zm19.5-4.5L23 13l-6.99 7-4.51-4.5L13 14l3.01 3 5.49-5.5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
